File: 프로그래머스/lv2/12909. 올바른 괄호/올바른 괄호.py
def solution(s):
    
    # 스택 활용
    
    stack = []
    
    # '('이면 무작정 스택에 넣고
    # ')'이면 스택 비어있는지 확인하고 있으면 pop
    # 비어있으면 바르게 짝지어지지 않음 return False
    # 복잡도 O(n)
    for i in s:
        if i == '(':
            stack.append(i)
        else :
            if stack:
                stack.pop()
            else:
                return False

    # 스택에 짝지어지지 않은 '('확인
    if stack:
        return False
    
    return True

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.Stack;

class Solution {
    boolean solution(String s) {
        
        Stack<String> stack = new Stack<>();
        
        for (int i = 0; i < s.length() ; i++) {
            String c = String.valueOf(s.charAt(i));
            
            if (c.equals("(")) {
                stack.push(c);
            } else {
                if ( !stack.isEmpty()) {
                    stack.pop();
                } else {
                    return false;
                }
            }
        }
        
        if (stack.isEmpty()) return true;
        else return false;
    }
}